src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/StartupObjectValueProvider.cs

```diff
@@ -0,0 +1,89 @@
+    private const string EnabledValue = "WindowsForms";
+    private const string DisabledValue = "WindowsFormsWithCustomSubMain";
+
+    internal const string UseWinFormsProperty = "UseWindowsForms";
+    internal const string StartupObjectProperty = "StartupObject";
+
+    [ImportingConstructor]
+    public StartupObjectValueProvider(IMyAppFileAccessor myAppXmlFileAccessor)
+    {
+        _myAppXmlFileAccessor = myAppXmlFileAccessor;
+    }
+
+    public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+    {
+        string isWindowsFormsProject = await defaultProperties.GetEvaluatedPropertyValueAsync(UseWinFormsProperty);
+
+        if (isWindowsFormsProject == "true")
+        {
+            string applicationFrameworkValue = await defaultProperties.GetEvaluatedPropertyValueAsync(ApplicationFrameworkProperty);
+
+            if (applicationFrameworkValue == EnabledValue)
+            {
+                if (unevaluatedPropertyValue.Contains("Form")) // Is there a better way to identify a form?
+                {
+                    // If the user selects a Form, the value should be serialized to the myapp file.
+                    await _myAppXmlFileAccessor.SetMainFormAsync(unevaluatedPropertyValue);
+                    await defaultProperties.DeletePropertyAsync(StartupObjectProperty);
+                    return null;
+                }
+
+                // If the ApplicationFramework is enabled, the value Sub Main should always be serialized to the project file.
+                await defaultProperties.SetPropertyValueAsync(StartupObjectProperty, "Sub Main");
+                await _myAppXmlFileAccessor.SetMainFormAsync(string.Empty);
+                return null;
+            }
+        }
+
+        // Else, if it's other than a Windows Forms project, save the StartupObject property in the project file as usual.
+        // Or if the ApplicationFramework property is disabled, save the StartupObject property in the project file as usual.
+        await defaultProperties.SetPropertyValueAsync(propertyName, unevaluatedPropertyValue);
+        return null;
+    }
+
+    public override async Task<string> OnGetEvaluatedPropertyValueAsync(string propertyName, string evaluatedPropertyValue, IProjectProperties defaultProperties)
+    {
+        // StartupObject can come from the project file or the myapp file.
+        string applicationFrameworkValue = await defaultProperties.GetEvaluatedPropertyValueAsync(ApplicationFrameworkProperty);
+        
+        if (applicationFrameworkValue == DisabledValue)
+            return await base.OnGetUnevaluatedPropertyValueAsync(propertyName, evaluatedPropertyValue, defaultProperties);
+
+        string valueInProjectFile = await base.OnGetUnevaluatedPropertyValueAsync(propertyName, evaluatedPropertyValue, defaultProperties);
```

Comment:
Would this be easier to read?

```suggestion
        string valueInProjectFile = await base.OnGetUnevaluatedPropertyValueAsync(propertyName, evaluatedPropertyValue, defaultProperties);

        if (applicationFrameworkValue == DisabledValue)
            return valueInProjectFile;
```

That diff's a bit of a mess. Basically, the local variable `valueInProjectFile` can be moved up, and its value re-used in the if block to reduce duplication. I found myself checking that both of the old calls were the same. Also "return valueInProjectFile" reads quite naturally.

Ditto below in the next method.